### DISCUSSION TITLE
MikroORM folder based discovery fails to persist entities using new Entity()

### DISCUSSION THREAD

**Discussion by kallekulp** [mikro-orm/mikro-orm#5804] [General]
### Describe the bug

I am getting the following error when trying to create any new entity with the constructor and trying to persisting it with `EntityManager` I get the following error:

`ValidationError: Trying to persist not discovered entity of type UserEntity. Entity with this name was discovered, but not the prototype you are passing to the ORM. If using EntitySchema, be sure to point to the implementation via `class`.`

I am not using `EntitySchema`. 

Based on the issue [here](https://github.com/mikro-orm/mikro-orm/issues/1131) I created a fresh app and the same error persist.

I am runnig the NX and MikroORM. 

Entities get discovered during app runtime/build. Running `npx mikro-orm debug` gives the following output:

```
Current MikroORM CLI configuration
 - dependencies:
   - mikro-orm 6.2.9
   - node 18.17.0
   - knex 3.1.0
   - mysql2 3.10.0
   - typescript 5.4.5
 - package.json found
 - ts-node enabled
 - searched config paths:
   - C:/Work/mikro-orm-nx/my-workspace/libs/database/src/lib/mikro-orm.config.ts (found)
   - C:/Work/mikro-orm-nx/my-workspace/dist/libs/database/src/lib/mikro-orm.config.js (found)
   - C:/Work/mikro-orm-nx/my-workspace/src/mikro-orm.config.ts (not found)
   - C:/Work/mikro-orm-nx/my-workspace/mikro-orm.config.ts (not found)
   - C:/Work/mikro-orm-nx/my-workspace/dist/mikro-orm.config.js (not found)
   - C:/Work/mikro-orm-nx/my-workspace/mikro-orm.config.js (not found)
 - configuration found
 - database connection successful
 - will use `entities` array (contains 0 references and 1 paths)
   - C:/Work/mikro-orm-nx/my-workspace/dist/**/entities/*.entity.js (found)
 - could use `entitiesTs` array (contains 0 references and 1 paths)
   - C:/Work/mikro-orm-nx/my-workspace/libs/database/src/lib/**/entities/*.entity.ts (found)
```

This is the config I am using:

```
const mikroOrmConfig = {
  user: process.env['DB_USERNAME'],
  host: process.env['DB_HOST'],
  password: process.env['DB_PASSWORD'],
  dbName: process.env['DB_NAME'],
  port: parseInt(process.env['DB_PORT'] as string),
  debug: true,
  driver: MySqlDriver,
  entities: [ 'dist/**/entities/*.entity.js' ],
  entitiesTs: [ 'libs/database/src/lib/**/entities/*.entity.ts' ],
  metadataProvider: ReflectMetadataProvider,
  extensions: [ SeedManager, Migrator ],
  migrations: {
    tableName: 'migration',
    path: './libs/database/src/lib/migrations',
    snapshot: true
  },
  seeder: {
    path: './seeders',
    pathTs: './seeders',
    defaultSeeder: 'DatabaseSeeder',
    glob: '!(*.d).{js,ts}',
    emit: 'ts',
    fileName: (className: string) => className
  }
} as MikroOrmModuleOptions;

export default mikroOrmConfig;
```

When console logging the following in the controller which should create the new user:

```
const meta = this.em.getMetadata().get(UserEntity);

console.log(meta.class);
console.log(UserEntity);
console.log(meta.class === UserEntity);
```

I see this:

```
[class UserEntity extends MyCustomBaseEntity]
[class UserEntity extends MyCustomBaseEntity]
false
```

What could be the issue here? Is something wrong with my config?

### Reproduction

Repository can be found [here](https://github.com/kallekulp/mikro-orm-nx)

Run `nx serve mikro-orm-nx` and open [http://localhost:3000/api](http://localhost:3000/api) from browser.

### What driver are you using?

@mikro-orm/mysql

### MikroORM version

6.2.9

### Node.js version

18.17.0

### Operating system

Windows

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by bigscaryduck** [comment]
Just confirming I was able to reproduce in the test app provided in the original ticket. 

Config looks correct to me, is this a bug?

**Comment by B4nan** [comment]
When you run the app, the ORM will try to detect whether you are using ts-node or some other way to bring in support for TS files, and based on that it will use either `entities` or `entitiesTs`. The error you see means you discovered one but used the other in your app most likely. This could be caused by the detection not working properly (which you can override with `tsNode: true` in the ORM config, but only for the dev runs, you don't want that in production).

I'll have a closer look at the repo and try to send you a PR fixing it.

**Reply by B4nan** [reply]
This looks quite suspicious, the compiled entity is there many times.

<img width="660" alt="image" src="https://github.com/mikro-orm/mikro-orm/assets/615580/159a1788-e148-4971-bbfb-1a0c687aa2b6">

**Reply by B4nan** [reply]
The problem seems to be about the fact that you run a compiled version of the app, which has the entities inlined. Those are surely different prototypes than the original compiled files.

<img width="691" alt="image" src="https://github.com/mikro-orm/mikro-orm/assets/615580/e55d63ff-890d-4b33-9332-6c46d3501dd1">

<img width="814" alt="image" src="https://github.com/mikro-orm/mikro-orm/assets/615580/05a92b7a-957f-4d57-9961-f5f13d63a5a6">

**Comment by B4nan** [comment]
Webpack and folder-based discovery is quite a shitshow, there are some notes on how it could be done here, mostly from the community, webpack was never my cup of tea.

https://mikro-orm.io/docs/next/deployment

I would rather discourage doing that, things will be much simpler if you do something like this instead:

```diff
import * as entities from './entities';

const mikroOrmConfig = {
+  entities: Object.values(entities),
-  entities: [ 'dist/**/entities/*.entity.js' ],
-  entitiesTs: [ 'libs/database/src/lib/**/entities/*.entity.ts' ],
```

This way the entities paths will be statically resolved by webpack, so things are resolved correctly to the bundled entities.

**Comment by bigscaryduck** [comment]
Thanks for the great detail and response!

**Comment by kallekulp** [comment]
Thank you for the response! So it's more of an NX issue than MikroORM. I will the entities way right now.

**Comment by kallekulp** [comment]
Also last question, if using the entities option in the config, do I still need to provide individual entities with `MikroOrmModule.forFeature({
			entities: [Physician, PhysicianMedicationPermission]
		})` in the feature module?

**Reply by B4nan** [reply]
Yes, that one is for registering repositories to nest DI container. But if you do that, you can as well use the `autoLoadEntities` option, which basically extracts the entity references from those `forFeature` calls and puts them into the ORM config automatically.

https://github.com/mikro-orm/nestjs#auto-entities-automatically

**Reply by B4nan** [reply]
Also, technically, you don't have to put all your entities to the ORM config, it will discover references automatically, e.g. if you have Book that points to Author entity, its enough to discover one (and use class references in the definition instead of strings). Similarly, base entities and embeddables are now also discovered automatically.

**Reply by kallekulp** [reply]
Thank you for the response again. The CLI will tell me no entities found when autoloading. I was using that before and the app was working fine but for testing it didn't, because I need to be able to tear down the database and construct again with the migration and then seed the data. meaning when I try to create initial migration it says no entities found unfortunately

**Reply by B4nan** [reply]
Yes, because the CLI doesn't know anything about your nest app.

Note that you can do everything programmatically too, and generally speaking, I would discourage using migrations for testing, I would just use schema generator directly.
